lib/ops/opsservice/update.go

```diff
@@ -330,23 +348,47 @@ func (s *site) createUpdateOperation(req ops.CreateSiteAppUpdateOperationRequest
 		return nil, trace.Wrap(err, "failed to create update operation")
 	}
 
-	resetSiteState := func() {
+	resetClusterState := func() {
 		if err == nil {
 			return
 		}
-		errReset := s.setSiteState(ops.SiteStateActive)
-		if errReset != nil {
-			log.Warningf("failed to reset site state: %v", trace.DebugReport(errReset))
+		logger := log.WithField("operation", op.Key())
+		// Fail the operation and reset cluster state.
+		// It is important to complete the operation as subsequent same type operations
+		// will not be able to complete if there's an existing incomplete one
+		if errReset := ops.FailOperationAndResetCluster(*key, s.service, err.Error()); errReset != nil {
+			logger.WithFields(log.Fields{
+				log.ErrorKey: errReset,
+				"operation":  key,
+			}).Warn("Failed to mark operation as failed.")
 		}
 	}
-	defer resetSiteState()
+	defer resetClusterState()
 
 	s.reportProgress(ctx, ops.ProgressEntry{
 		State:      ops.ProgressStateInProgress,
 		Completion: 0,
 		Message:    "initializing the operation",
 	})
 
+	if !req.StartAgents {
+		return key, nil
+	}
+
+	updatePackage, err := loc.ParseLocator(req.App)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	updateApp, err := s.apps().GetApp(*updatePackage)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	err = s.startUpdateAgent(context.TODO(), ctx, updateApp)
```

Comment:
This is the forward port of the issue where the logic to support updates triggered from UI was removed by mistake.